### PR TITLE
Add finalizer cleanup logic to service controller to support lb finalizer protection

### PR DESCRIPTION
**Pull request by jhorwit2**
**What this PR does / why we need it**:

This PR adds finalizer cleanup logic to the service controller which will enable a future release to add finalizer logic to the service controller that ensures that the cloud resources are cleaned up prior to removing the finalizer. This ensures that a rollback to the prior version of Kubernetes does not break services with the finalizer added. 

Ref #54569

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add support for cleaning up service load balancer finalizers to enable rollbacks once we release finalizer protection when deleting load balancer services. 
```

### CODE REVIEW

File: pkg/util/service/service.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package service
```
Comment (neolit123) [inline pkg/util/service/service.go]: tests are complaining about a missing b-plate header.

File: pkg/util/service/service.go [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+const (
+	LoadBalancerProtectionFinalizer = "kubernetes.io/load-balancer-service-protection"
```
Comment (neolit123) [inline pkg/util/service/service.go]: should be commented as exported.

File: pkg/controller/service/service_controller.go [hunk 3]
```diff
@@ -282,6 +291,7 @@ func (s *ServiceController) createLoadBalancerIfNeeded(key string, service *v1.S
 	var err error
 
 	if !wantsLoadBalancer(service) {
+
```
Comment (neolit123) [inline pkg/controller/service/service_controller.go]: there is some stray whitespace in the diff, but this is just a nit. :)

**Comment by feiskyer** [comment]
@jhorwit2 Any updates on this feature?

File: pkg/util/service/service.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+const (
+	// LoadBalancerProtectionFinalizer is the finalizer added to load balancer services to ensure
+	// that after successful creation the cloud resources are removed before deleting the service.
+	LoadBalancerProtectionFinalizer = "kubernetes.io/load-balancer-service-protection"
```
Comment (andrewsykim) [inline pkg/util/service/service.go]: Thoughts on using `networking.kubernetes.io/load-balancer-protection` instead?

Comment (andrewsykim) [reply pkg/util/service/service.go]: or `networking.kubernetes.io/load-balancer-finalizer`

File: pkg/util/service/service.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
+/*
```
Comment (andrewsykim) [inline pkg/util/service/service.go]: Thoughts on moving this to https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/cloud-provider/service?

File: pkg/util/service/service.go (hunk 4 above)
Comment (jhorwit2) [reply pkg/util/service/service.go]: @andrewsykim I picked this name so there was consistency with the PV protection.
https://github.com/kubernetes/kubernetes/blob/b11211ed8cfbf5b613c9dfb20affa6627c5e7ac7/pkg/volume/util/finalizer.go#L24

File: pkg/util/service/service.go (hunk 5 above)
Comment (jhorwit2) [reply pkg/util/service/service.go]: We could but this is a general util file so putting it there could eventually mean duplication of code. The reason I put it here is there are similar utilities in the util package for other resources.

**Comment by MrHohn** [comment]
Closing this as https://github.com/kubernetes/kubernetes/pull/78262 was merged.
/close

**Pull request closed without merging**